{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Invitae_Bioinformatics_Exercise_B.3\n",
    "<hr>\n",
    "By: Barry Guglielmo, ALM\n",
    "<br>\n",
    "Date: 10/8/2021\n",
    "<hr>\n",
    "\n",
    "### Objective\n",
    "This code was written in pursute of the Bioinfomatics Engineer, Pipeline position at Invitae.<br>\n",
    "The objective is to write software that translates transcript coordinates to genomic coordinates.\n",
    "<br><br>\n",
    "\n",
    "### Links\n",
    "<ul>\n",
    "<li><a href = \"Invitae_Bioinformatics_Exercise_B.3.pdf\">Exercise PDF</a></li>\n",
    "<li><a href = \"https://www.datahound.org\">Personal Portfolio (Python Flask Hosted on Azure)</a></li>\n",
    "<li><a href = \"\">GitHub</a></li>\n",
    "<li><a href = \"\">LinkedIn</a></li>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "class cigar():\n",
    "    '''translates transcript coordinates to genomic coordinates.\n",
    "    \n",
    "        Keyword arguments:\n",
    "        t1 -- path to input cigar strings\n",
    "        t2 -- list of queries to use\n",
    "        outpath -- output path for file to be written, must end in / or be blank for local directory\n",
    "        outname -- name of the output file\n",
    "        \n",
    "        Warning:\n",
    "        Queries longer than the cigar string will be treated as matches\n",
    "        \n",
    "    '''\n",
    "    \n",
    "    def __init__(self, t1, t2, write = True, outpath = \"\", outname = \"transcript_genome_coordinates.tsv\"):\n",
    "        self.t1 = pd.read_csv(t1, sep = \"\\t\", header = None) # input sequences\n",
    "        self.t2 = pd.read_csv(t2, sep = \"\\t\", header = None) # queries\n",
    "        self.tg = [] # array of outputs to be transformed to pandas dataframe at end\n",
    "        for index, row in self.t2.iterrows():# go through t2 (queries)\n",
    "            loc = self.t1.loc[self.t1[0]==row[0]].iloc[0] # assuming that each is only used once\n",
    "            p = re.findall(r'(\\d+)([A-Z]{1})', loc[3]) # parse cigar string \n",
    "            c = row[1] # transcript coordinate as object to be manipulated without damaging reference\n",
    "            if c<0: # error handler for negative entries\n",
    "                print(\"Warning: Alignment should stat at 0 or positive integer\")\n",
    "            skip = False # a handler if the start point is 0\n",
    "            for i in range(0,len(p)): # find index of the end of the transcript\n",
    "                if i == 0 and c == 0: # if the start point is 0 go no further\n",
    "                    e = i\n",
    "                    skip = True\n",
    "                    break\n",
    "                elif c>=0: # else count down matches and insertions seen (deletions ignored and index will increase)\n",
    "                    if p[i][1]=='M':\n",
    "                        c-=int(p[i][0])\n",
    "                    if p[i][1]=='I':\n",
    "                        c-=(int(p[i][0]))\n",
    "                        if c<0:\n",
    "                            c = -1 # insertions not cumulative\n",
    "                    if c<=0:\n",
    "                        e = i\n",
    "                else:\n",
    "                    break \n",
    "            f = int(loc[2]) # final genome location initialized as the location of start of alignment\n",
    "            if skip!=True:  # if start point not 0 add to genome location, otherwise skip\n",
    "                for i in range(0,e+1):  # use the index to calculate the genome location\n",
    "                    if p[i][1]=='M' or p[i][1]=='D': # matches and deletions counted (insertions ignored but index will increase)\n",
    "                        f+=int(p[i][0])\n",
    "                f+=c # add the remainder from calcualting transcript end index (-) to genome location\n",
    "            self.tg.append([row[0],row[1],loc[1],f])# make list that is now accessible by the class\n",
    "        self.tg = pd.DataFrame(self.tg) # transform to pandas data frame for ease of use\n",
    "        if write == True: # write to outfile\n",
    "            try:\n",
    "                self.tg.to_csv(\"%s%s\"%(outpath, outname),sep = '\\t', index = False, header = False )\n",
    "            except:\n",
    "                print(\"Outpath must end in / or be blank for current directory\")\n",
    "    \n",
    "    def compare(self, standard):\n",
    "        '''Do a comparison of self.tg with a known correct output file\n",
    "        \n",
    "            Keyword arguments:\n",
    "            standard -- path to a file with known correct outputs in same format as output from cigar\n",
    "        '''\n",
    "        self.standard = pd.read_csv(standard, sep = \"\\t\", header = None)\n",
    "        outcome = []\n",
    "        for index, row in self.standard.iterrows():\n",
    "            d = self.tg.loc[(self.tg[0]==row[0])&(self.tg[1]==row[1])]\n",
    "            try:\n",
    "                if list(d.iloc[0]) == list(row):\n",
    "                    outcome.append(\"correct,%s,%s\"%(list(row),list(d.iloc[0])))\n",
    "                else:\n",
    "                    outcome.append(\"wrong,%s,%s\"%(list(row),list(d.iloc[0])))\n",
    "            except:\n",
    "                print(\"Out of Bounds,%s,%s\"%(list(row),\"No Matching Transcript\"))\n",
    "        return outcome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "correct,['TR1', 4, 'CHR1', 7],['TR1', 4, 'CHR1', 7]\n",
      "correct,['TR1', 7, 'CHR1', 10],['TR1', 7, 'CHR1', 10]\n",
      "correct,['TR1', 8, 'CHR1', 18],['TR1', 8, 'CHR1', 18]\n",
      "correct,['TR1', 9, 'CHR1', 19],['TR1', 9, 'CHR1', 19]\n",
      "correct,['TR1', 12, 'CHR1', 22],['TR1', 12, 'CHR1', 22]\n",
      "correct,['TR1', 13, 'CHR1', 23],['TR1', 13, 'CHR1', 23]\n",
      "correct,['TR1', 14, 'CHR1', 23],['TR1', 14, 'CHR1', 23]\n",
      "correct,['TR1', 15, 'CHR1', 23],['TR1', 15, 'CHR1', 23]\n",
      "correct,['TR1', 16, 'CHR1', 24],['TR1', 16, 'CHR1', 24]\n",
      "2\n",
      "correct,['TR2', 0, 'CHR2', 10],['TR2', 0, 'CHR2', 10]\n",
      "correct,['TR2', 7, 'CHR2', 17],['TR2', 7, 'CHR2', 17]\n",
      "3\n",
      "correct,['TR3', 2, 'CHR3', 5],['TR3', 2, 'CHR3', 5]\n",
      "correct,['TR3', 3, 'CHR3', 5],['TR3', 3, 'CHR3', 5]\n",
      "correct,['TR3', 4, 'CHR3', 5],['TR3', 4, 'CHR3', 5]\n",
      "correct,['TR3', 5, 'CHR3', 6],['TR3', 5, 'CHR3', 6]\n",
      "correct,['TR3', 6, 'CHR3', 7],['TR3', 6, 'CHR3', 7]\n",
      "4\n",
      "correct,['TR4', 2, 'CHR4', 5],['TR4', 2, 'CHR4', 5]\n",
      "correct,['TR4', 3, 'CHR4', 5],['TR4', 3, 'CHR4', 5]\n",
      "correct,['TR4', 4, 'CHR4', 5],['TR4', 4, 'CHR4', 5]\n",
      "correct,['TR4', 5, 'CHR4', 8],['TR4', 5, 'CHR4', 8]\n",
      "correct,['TR4', 6, 'CHR4', 9],['TR4', 6, 'CHR4', 9]\n",
      "correct,['TR4', 7, 'CHR4', 10],['TR4', 7, 'CHR4', 10]\n",
      "5\n",
      "correct,['TR5', 0, 'CHR5', 3],['TR5', 0, 'CHR5', 3]\n",
      "correct,['TR5', 2, 'CHR5', 5],['TR5', 2, 'CHR5', 5]\n",
      "correct,['TR5', 3, 'CHR5', 7],['TR5', 3, 'CHR5', 7]\n",
      "correct,['TR5', 4, 'CHR5', 7],['TR5', 4, 'CHR5', 7]\n",
      "correct,['TR5', 5, 'CHR5', 8],['TR5', 5, 'CHR5', 8]\n",
      "correct,['TR5', 6, 'CHR5', 9],['TR5', 6, 'CHR5', 9]\n",
      "correct,['TR5', 7, 'CHR5', 10],['TR5', 7, 'CHR5', 10]\n"
     ]
    }
   ],
   "source": [
    "for i in range(1,6):\n",
    "    print(i)\n",
    "    c = cigar(\"files/t1.tsv\",\"files/tr%i.tsv\"%i)\n",
    "    s = c.compare(\"files/tr%icorrect.tsv\"%i)\n",
    "    [print(j) for j in s]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
